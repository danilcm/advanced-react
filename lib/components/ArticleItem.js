import React from 'react';

import PropTypes from 'prop-types';
import storeProvider from './storeProvider';

const styles = {
  main: {
    marginBottom: 50
  },
  title: {
    fontFamily: 'Product Sans',
    fontWeight: 'bold'
  },
  link: {
    fontFamily: 'Product Sans',
    fontWeight: 'bold',
    fontSize: 15,
    color: '#1b649f',
    textDecoration: 'none',
  },
  body: {
    maxWidth: 400,
    textAlign: 'justify',
    fontFamily: 'Georgia'
  },
  date: {
    fontSize: 14,
    fontFamily: 'Product Sans',
    color: '#666666'
  }
};

const prettyDate = (isoString) => new Date(isoString).toDateString();

const ArticleItem = ({ author, article }) => {
  return (
    <div style={styles.main}>
      <h3 field="title" style={styles.title}>{article.title}</h3>
      <a field="website" style={styles.link} href={author.website}>{`${author.firstName} ${author.lastName}`}</a>
      <p field="body" style={styles.body}>{article.body}</p>
      <span field="date" style={styles.date}>{prettyDate(article.date)}</span>
    </div>
  );
};

ArticleItem.propTypes = {
  article: PropTypes.shape({
    body: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  }),
  author: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    website: PropTypes.string.isRequired,
  })
};

ArticleItem.displayName = 'Article';

const mapExtraProps = (store, props) => ({
  author: store.getAuthor(props.article.authorId)
});

export const ArticleItemPure = ArticleItem;

export default storeProvider(mapExtraProps)(ArticleItem);

