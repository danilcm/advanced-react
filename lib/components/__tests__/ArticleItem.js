import React from 'react';

import Enzyme, { shallow } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';

import toJson from 'enzyme-to-json';

// Setup enzyme's react adapter
Enzyme.configure({ adapter: new EnzymeAdapter() });

import { ArticleItemPure } from '../ArticleItem';

const testProps = {
  article: {
    'id': '95c12a8f6c88953ca8f8a39da25546e6',
    'title': 'Introducing React\'s Error Code System',
    'date': 'Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)',
    'authorId': '2c6aa2cfe3449467d329fa17d6ea230f',
    'body': 'wire.'
  },
  author: {
    'id': '335fb02ec8f76c8515821ac9f266d276',
    'firstName': 'Joseph',
    'lastName': 'Savona',
    'website': 'https://twitter.com/en_JS'
  }
};

const setup = () => shallow(<ArticleItemPure {...testProps}/>);

const checkByField = (field, expected, additional = () => {}) => (tree) => {
  const node = tree.find(`[field="${field}"]`);
  
  expect(node.text()).toBe(expected);
  additional(node);
};

describe('ArticleItem', () => {
  const tree = setup();

  it('should match the snapshot', () => {
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should render correct data', () => {
    checkByField(
      'title', 
      testProps.article.title
    )(tree);
    
    checkByField(
      'website', 
      `${testProps.author.firstName} ${testProps.author.lastName}`,
      (node) => expect(node.prop('href')).toBe(testProps.author.website)
    )(tree);
    
    checkByField(
      'body', 
      testProps.article.body
    )(tree);

    checkByField(
      'date', 
      new Date(testProps.article.date).toDateString()
    )(tree);
  });
  
});


