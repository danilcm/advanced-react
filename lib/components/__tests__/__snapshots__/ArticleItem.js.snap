// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleItem should match the snapshot 1`] = `
<div
  style={
    Object {
      "marginBottom": 50,
    }
  }
>
  <h3
    field="title"
    style={
      Object {
        "fontFamily": "Product Sans",
        "fontWeight": "bold",
      }
    }
  >
    Introducing React's Error Code System
  </h3>
  <a
    field="website"
    href="https://twitter.com/en_JS"
    style={
      Object {
        "color": "#1b649f",
        "fontFamily": "Product Sans",
        "fontSize": 15,
        "fontWeight": "bold",
        "textDecoration": "none",
      }
    }
  >
    Joseph Savona
  </a>
  <p
    field="body"
    style={
      Object {
        "fontFamily": "Georgia",
        "maxWidth": 400,
        "textAlign": "justify",
      }
    }
  >
    wire.
  </p>
  <span
    field="date"
    style={
      Object {
        "color": "#666666",
        "fontFamily": "Product Sans",
        "fontSize": 14,
      }
    }
  >
    Mon Jul 11 2016
  </span>
</div>
`;
